### PR TITLE
Feature/3863 permissive source routing

### LINKED ISSUE #3863: Permissive source routing
## Abstract

We need a specification for permissive source routing every important stakeholder agrees on. Please update the Raiden spec.

Due date: End of Iteration "Spring Cleaning" (03.05.2019)

## Motivation

Lower the cost of mediation by reusing the path from the initiator.

## Specification

- Drop support to the UDP transport (it does not support variable length messages, ATM) #3894
- Implemented Payment metadata #1202
- Add the path as metadata from the initiator
- Initially mediator should fail if matching proposed routes can not be used
  - This should be improved, the mediator can query the PFS again if the route can't be used.
- Mediators should query the PFS if the message does not contain any routing metadata
- The hops should not trim the paths

## Backwards Compatibility

This change is backwards compatible, any compatibility problems are introduced/handled by #1202

### LINKED ISSUE #4142: Don't request new routes in `handle_message_refundtransfer`
Currently `handle_message_refundtransfer` requests new routes from the PFS.

https://github.com/raiden-network/raiden/blob/fca59bd3e632d33cf26d74583380bcfef468e037/raiden/message_handler.py#L117-L125

Instead it should use the stored routes in `InitiatorPaymentState`.

### PR DESCRIPTION
**Pull request by lullis**
Fixes #3863 
Resolves #4142 

Introduces the concept of permissive source routing, where the initiator nodes send not one single route for mediators, but a list of candidate routes. This information is sent/managed by the `RouteState` type, which defines an ordered list of addresses (from starting point to end point) and an optional `forward_channel_id`, which can be used as a hint for which hop to take.

Also adds new container fields for Transfer-related messages: `Metadata` contains a list of `RouteMetadata` fields, where each RouteMetadata is a simple list of addresses.

### CODE REVIEW

File: raiden/transfer/mediated_transfer/initiator.py [hunk 1]
```diff
@@ -179,21 +179,40 @@ def get_initial_lock_expiration(
...
+        candidate_channel_state = forward_channel_id and channelidentifiers_to_channels.get(
+            forward_channel_id
+        )
+
+        if (
+            forward_channel_id
+            and candidate_channel_state
+            and channel.is_channel_usable(
+                candidate_channel_state=candidate_channel_state, transfer_amount=amount_with_fee
+            )
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/initiator.py]: All these conditions could become:
```
is_usable_route = (
            forward_channel_id
            and candidate_channel_state
            and channel.is_channel_usable(
                candidate_channel_state=candidate_channel_state, transfer_amount=amount_with_fee
            )
if is_usable_route:
 ....
```

File: raiden/transfer/mediated_transfer/mediator.py [hunk 2]
```diff
@@ -354,35 +354,35 @@ def clear_if_finalized(
...
     block_number: BlockNumber,
 ) -> Tuple[Optional[MediationPairState], List[Event]]:
-    """ Given a payer transfer tries a new route to proceed with the mediation.
+    """ Given a payer transfer tries the given route to proceed with the mediation.
 
     Args:
         payer_transfer: The transfer received from the payer_channel.
-        available_routes: Current available routes that may be used, it's
-            assumed that the routes list is ordered from best to worst.
+        route_state: oute to be tried.
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/mediator.py]:
```suggestion
        route_state: route to be tried.
```

File: raiden/transfer/mediated_transfer/mediator.py [hunk 3]
```diff
@@ -996,18 +1003,28 @@ def mediate_transfer(
...
-        available_routes,
-        channelidentifiers_to_channels,
-        pseudo_random_generator,
-        block_number,
-    )
+    transfer_pair: Optional[MediationPairState] = None
+    mediated_events: List[Event] = list()
+
+    reachable_route_states = [
+        r
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/mediator.py]: please use a complete meaningful name

File: raiden/transfer/mediated_transfer/mediator.py [hunk 4]
```diff
@@ -354,35 +354,35 @@ def clear_if_finalized(
...
+
+    if payee_channel and channel.is_channel_usable(
+        candidate_channel_state=payee_channel,
         transfer_amount=payer_transfer.lock.amount,
         lock_timeout=lock_timeout,
-    )
-
-    if route_infos:
-        payee_channel, route_state = route_infos
+    ):
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/mediator.py]: Save the predicate result into a variable and use that in the if-statement

File: raiden/transfer/mediated_transfer/mediator.py [hunk 5]
```diff
@@ -996,18 +1003,28 @@ def mediate_transfer(
...
-        payer_transfer,
-        available_routes,
-        channelidentifiers_to_channels,
-        pseudo_random_generator,
-        block_number,
-    )
+    transfer_pair: Optional[MediationPairState] = None
+    mediated_events: List[Event] = list()
+
+    reachable_route_states = [
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/mediator.py]: What you've done here is remove a call to `filter_reachable_routes` which originally filtered the routes according to the availability of the the nodes and manually did the same thing here. Let's please maintain one function which is preferrable the one you removed and do the filtering there.

File: raiden/transfer/mediated_transfer/mediator.py [hunk 6]
```diff
@@ -996,18 +1003,28 @@ def mediate_transfer(
     send a refund back to the payer, allowing the payer to try a different
     route.
     """
-    reachable_routes = filter_reachable_routes(possible_routes, nodeaddresses_to_networkstates)
-    available_routes = filter_used_routes(state.transfers_pair, reachable_routes)
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/mediator.py]: Why did you remove the call to this function?

File: raiden/transfer/mediated_transfer/mediator.py (hunk 4 above)
Comment (rakanalh) [reply raiden/transfer/mediated_transfer/mediator.py]: Actually, @hackaugusto and I seem to agree that all filtering should be moved to the caller of this function so that all filtering happens in one place.

File: raiden/raiden_service.py [hunk 7]
```diff
@@ -163,28 +163,26 @@ def initiator_init(
     return ActionInitInitiator(transfer_state, routes)
 
 
-def mediator_init(raiden: "RaidenService", transfer: LockedTransfer) -> ActionInitMediator:
+def mediator_init(
+    raiden: "RaidenService", transfer: LockedTransfer
+) -> Optional[ActionInitMediator]:
     from_transfer = lockedtransfersigned_from_message(transfer)
     # Feedback token not used here, will be removed with source routing
```
Comment (palango) [inline raiden/raiden_service.py]: This comment can go as well

File: raiden/routing.py [hunk 8]
```diff
@@ -281,3 +282,32 @@ def get_best_routes_pfs(
...
+        if channel_state is not None:
+            resolvable.append(
+                RouteState(
+                    route=route_metadata.route,
+                    forward_channel_id=channel_state.canonical_identifier.channel_identifier,
+                )
+            )
+        else:
+            unresolvable.append(RouteState(route=route_metadata.route))
+    return resolvable + unresolvable
```
[the remark this answers is not in the document]
Comment (palango) [reply raiden/routing.py]: Please create a tracking issue so we keep this in mind and can change/fix it in the future.

File: raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate.py [hunk 9]
```diff
@@ -156,61 +157,21 @@ def test_next_route_amount():
...
-                canonical_identifier=identifiers[3], reveal_timeout=timeout_blocks
-            ),
-        ]
-    )
-
-    chosen_channel, _ = channel.next_channel_from_routes(
-        channels.get_routes(0, 1, 2, 3), channels.channel_map, UNIT_TRANSFER_AMOUNT, timeout_blocks
-    )
-    assert chosen_channel.identifier == channels[2].identifier
+    # channels
```
Comment (palango) [inline raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate.py]: I guess there should be more in this comment

File: raiden/tests/unit/transfer/test_source_routing.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        token_network_address=token_network_state.address,
+        chain_state=chain_state,
+    )
+
+    msg = "route resolved with wrong channel id"
+    channel_id = netting_channel_state.canonical_identifier.channel_identifier
+    assert route_states[0].forward_channel_id == channel_id, msg
+
+
+def test_mediator_forwards_pruned_route():
```
Comment (palango) [inline raiden/tests/unit/transfer/test_source_routing.py]: This should be removed or do something

File: raiden/transfer/channel.py [hunk 11]
```diff
@@ -1256,6 +1239,7 @@ def send_lockedtransfer(
     payment_identifier: PaymentID,
     expiration: BlockExpiration,
     secrethash: SecretHash,
+    route_states: List[RouteState],
 ) -> SendLockedTransfer:
     send_locked_transfer_event, merkletree = create_sendlockedtransfer(
         channel_state,
```
Comment (palango) [inline raiden/transfer/channel.py]: Can you add kwargs here as well

File: raiden/message_handler.py [hunk 12]
```diff
@@ -110,35 +114,25 @@ def handle_message_lockexpired(raiden: RaidenService, message: LockExpired) -> N
...
         )
 
-        if role == "initiator":
-            old_secret = views.get_transfer_secret(chain_state, from_transfer.lock.secrethash)
-            # We currently don't allow multi routes if the initiator does not
-            # hold the secret. In such case we remove all other possible routes
-            # which allow the API call to return with with an error message.
-            if old_secret == ABSENT_SECRET:
-                routes = list()
+        routes: List[RouteState] = list()
```
[the remark this answers is not in the document]
Comment (lullis) [reply raiden/message_handler.py]: Done

File: raiden/message_handler.py [hunk 13]
```diff
@@ -110,35 +114,25 @@ def handle_message_lockexpired(raiden: RaidenService, message: LockExpired) -> N
...
         role = views.get_transfer_role(
             chain_state=chain_state, secrethash=from_transfer.lock.secrethash
         )
 
-        if role == "initiator":
-            old_secret = views.get_transfer_secret(chain_state, from_transfer.lock.secrethash)
-            # We currently don't allow multi routes if the initiator does not
-            # hold the secret. In such case we remove all other possible routes
-            # which allow the API call to return with with an error message.
-            if old_secret == ABSENT_SECRET:
```
[the remark this answers is not in the document]
Comment (lullis) [reply raiden/message_handler.py]: Please check it again. It seems that now we do the check for atomic swaps and do not allow retries (by setting `is_reroute_allowed` to `False`) when there is a secret.

File: raiden/routing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lullis) [reply raiden/routing.py]: I believe the idea behind it was to sort all routes by availability/priority, but still keeps routes that are unresolvable at one point but could still come back online.

[the remark this answers is not in the document]
Comment (palango) [reply raiden/routing.py]: > I believe the idea behind it was to sort all routes by availability/priority, but still keeps routes that are unresolvable at one point but could still come back online.

I don't see a check about availability here, it only checks whether or not a channel to the partner exists.

[the remark this answers is not in the document]
Comment (lullis) [reply raiden/routing.py]: That is true. But to be honest I don't know if we have any spec'ed behavior for this case. Would it be okay to add this as an issue to #4236 ?

[the remark this answers is not in the document]
Comment (palango) [reply raiden/routing.py]: Sure!

File: raiden/transfer/state.py [hunk 14]
```diff
@@ -186,13 +186,18 @@ class RouteState(State):
 
     # TODO: Add timestamp
     route: List[Address]
-    forward_channel_id: ChannelID
+    forward_channel_id: Optional[ChannelID] = None
```
Comment (hackaugusto) [inline raiden/transfer/state.py]: I don't think this should be made `Optional`

File: raiden/transfer/mediated_transfer/mediator.py [hunk 15]
```diff
@@ -354,35 +348,41 @@ def clear_if_finalized(
...
+
+    is_payee_channel_usable = payee_channel is not None and channel.is_channel_usable(
+        candidate_channel_state=payee_channel,
         transfer_amount=payer_transfer.lock.amount,
         lock_timeout=lock_timeout,
     )
 
-    if route_infos:
-        payee_channel, route_state = route_infos
+    if payee_channel is not None and is_payee_channel_usable:
```
Comment (hackaugusto) [inline raiden/transfer/mediated_transfer/mediator.py]: This doesn't look right. What if one of the other routes in `route_state_table` can be used?

File: raiden/transfer/mediated_transfer/mediator.py [hunk 16]
```diff
@@ -445,6 +445,12 @@ def backward_transfer_pair(
     # do anything and wait for the received lock to expire.
     if channel.is_channel_usable(backward_channel, lock.amount, lock_timeout):
         message_identifier = message_identifier_from_prng(pseudo_random_generator)
+
+        backward_route_state = RouteState(
```
Comment (hackaugusto) [inline raiden/transfer/mediated_transfer/mediator.py]: I think we should not send a route backwards

File: raiden/transfer/channel.py [hunk 17]
```diff
@@ -1408,6 +1392,7 @@ def send_refundtransfer(
     payment_identifier: PaymentID,
     expiration: BlockExpiration,
     secrethash: SecretHash,
+    route_state: RouteState,
```
Comment (hackaugusto) [inline raiden/transfer/channel.py]: I think it makes more sense to *not* send a route_state backwards.

File: raiden/transfer/mediated_transfer/mediator.py [hunk 18]
```diff
@@ -996,19 +1001,27 @@ def mediate_transfer(
...
-        pseudo_random_generator,
-        block_number,
+    transfer_pair: Optional[MediationPairState] = None
+    mediated_events: List[Event] = list()
+
+    reachable_route_states = filter_reachable_routes(
+        candidate_route_states, nodeaddresses_to_networkstates
     )
 
+    for route_state in reachable_route_states:
```
Comment (hackaugusto) [inline raiden/transfer/mediated_transfer/mediator.py]: Could you please move the route filtering to a single function? It is important to call ` filter_reachable_routes`, `filter_used_routes`, and `channel.is_channel_usable`, having them together should it make it easier to review.

File: raiden/transfer/mediated_transfer/mediator.py [hunk 19]
```diff
@@ -1165,15 +1179,26 @@ def handle_refundtransfer(
         if not is_valid:
             return TransitionResult(mediator_state, channel_events)
 
+        # We remove all routes where the sender of the refund is our forward hop
```
Comment (hackaugusto) [inline raiden/transfer/mediated_transfer/mediator.py]: This is what `filter_used_routes` does (or is suppose to do), please use that function.

File: raiden/transfer/mediated_transfer/mediator.py [hunk 20]
```diff
@@ -1378,6 +1403,8 @@ def handle_node_change_network_state(
     if mediator_state.waiting_transfer is None:
         return TransitionResult(mediator_state, list())
 
+    assert route.forward_channel_id
```
Comment (hackaugusto) [inline raiden/transfer/mediated_transfer/mediator.py]: Please make this non optional

File: raiden/transfer/mediated_transfer/mediator.py [hunk 21]
```diff
@@ -980,7 +985,7 @@ def secret_learned(
 
 def mediate_transfer(
     state: MediatorTransferState,
-    possible_routes: List[RouteState],
+    candidate_route_states: List[RouteState],
```
Comment (hackaugusto) [inline raiden/transfer/mediated_transfer/mediator.py]: Please, persist this value, it then can be used on a refund (if there are more then one route in there)

File: raiden/transfer/state.py [hunk 22]
```diff
@@ -186,13 +186,18 @@ class RouteState(State):
...
 
     @property
     def next_hop_address(self) -> Address:
-        assert len(self.route) >= 2
+        assert len(self.route) >= 1
         return self.route[1]
 
+    def __repr__(self):
+        return "RouteState ({}), channel_id: {}".format(
+            " -> ".join([pex(a) for a in self.route]), self.forward_channel_id
```
Comment (hackaugusto) [inline raiden/transfer/state.py]: please use `to_checksumed_address` instead.

Comment (hackaugusto) [reply raiden/transfer/state.py]: also, the inner list is not necessary, a generator will suffice

File: raiden/transfer/mediated_transfer/mediator.py (hunk 18 above)
Comment (lullis) [reply raiden/transfer/mediated_transfer/mediator.py]: I am not sure I follow. The idea would be to have one "master" function that takes all the logic related to these functions? At the moment, the check for "is_channel_usable" is only at the mediator part.

**Comment by lullis** [comment]
@rakanalh, I addressed the changes that @hackaugusto requested. The part about `filter_reachable_routes` and `filter_used_routes` being placed back as its own functions will be addressed in a different issue which I was already working on.

File: raiden/message_handler.py [hunk 23]
```diff
@@ -160,49 +159,29 @@ def handle_message_lockexpired(raiden: RaidenService, message: LockExpired) -> N
...
             secret = random_secret()
+            old_secret = views.get_transfer_secret(chain_state, from_transfer.lock.secrethash)
+            is_reroute_allowed = bool(old_secret) and old_secret != ABSENT_SECRET
             state_change: StateChange = ReceiveTransferRefundCancelRoute(
-                routes=routes,
                 transfer=from_transfer,
                 balance_proof=from_transfer.balance_proof,
                 sender=from_transfer.balance_proof.sender,  # pylint: disable=no-member
                 secret=secret,
+                is_reroute_allowed=is_reroute_allowed,
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/message_handler.py]: Issue #4257
